### DISCUSSION TITLE
close multiple drives at once

### DISCUSSION THREAD

**Discussion by totoantibes** [teslamate-org/teslamate#2364] [Ideas]
I have to regularly  (and i mean daily) go on teslamate and look for missing drives in grafana, look at the missing ID number from the overlay url and then run the "close drive script" manually for the missing ID. is there by any chance a small script that could run a "for loop" to clean it up? i could cron the script and have a clean up state of trips.... It should not be that difficult but i am not sure what is the query needed to look for the missing trips...

this is what i run manually (as per the docs: [https://docs.teslamate.org/docs/maintenance/manually_fixing_data](url)
`sudo docker-compose exec teslamate bin/teslamate rpc     "TeslaMate.Repo.get!(TeslaMate.Log.Drive,1125) |> TeslaMate.Log.close_drive()"`

**Comment by init6-git** [comment]
It's been a long time since I wrote a shell script, but just knocked this one up:

```
#!/bin/bash
echo "Looking for unclosed drives"
records=$(cat ./unclosed-drives.sql | docker exec -i CONTAINER_NAME psql -t --pset="footer=off" -U teslamate -d teslamate)

echo "Closing drives"

for x in $records; 

do docker-compose exec teslamate bin/teslamate rpc "TeslaMate.Repo.get(TeslaMate.Log.Drive, $x) |> TeslaMate.Log.close_drive()" ; 
        echo "drive $x closed"
done;

echo "All closed"
```

Requires a file called unclosed-drives.sql which contains:
`select id from drives where end_date is null order by start_date;`


I'm sure these should be put into one script, but I've done enough today :)

**Reply by totoantibes** [reply]
thanks for your efforts! I have since then realised that my enclosed drives were due to the default timeout on the grafana DB. so i extended the timeout and all my drives are now closed without needing to do housecleaning as requested above.

**Reply by BabyYeggie** [reply]
What did you extend the timeout to? 90 seconds?

**Reply by totoantibes** [reply]
240000 (4 mins)  no issues si far. I think that the longer teslamate has been installed the longer the queries are. My install is now 2 years old with 35000kmin the belly.

**Reply by jeallen65** [reply]
How did you extend the timeout in Grafana? I’m having the same issue.

**Reply by totoantibes** [reply]
check out this page: [https://docs.teslamate.org/docs/configuration/environment_variables](https://docs.teslamate.org/docs/configuration/environment_variables)

you will find the DATABASE_TIMEOUT param.

**Reply by jeallen65** [reply]
That did it! Thanks so much for your help!!

**Reply by chinedu40** [reply]
@totoantibes Thank you for mentioning this fix to the issue, it had been causing problems for me for over 2 years and changing the database timeout has fixed the issue for me. 

@adriankumpf perhaps a change to the default database timeout should be considered and/or make this something you put in the wiki as a fix?

**Reply by aiden-cat** [reply]
For those curious, this can dump the outputs into the necessary file:

`docker exec pi_database_1 psql -d teslamate -U teslamate -c "select id from drives where end_date is null order by start_date;" -e > unclosed-drives.sql`

**Reply by dmachadol** [reply]
I created this script and it works perfectly:

- Searches for unclosed charges and drives records that are 2 days old or more (This prevents active charge/drive from closing)

- You can add it to the system cron or run it manually at your convenience.

```
#!/bin/bash
echo "Seeking for unclosed drives & charges..."

sql="select id from charging_processes where end_date is null and start_date <  now() + interval '-2 days' order by start_date;"
data=$(docker exec teslamate_database_1 bash -c "psql -d teslamate -U teslamate -t -c \"${sql}\"")
data=($data)

if [ "${data}" != "" ]; then
  for m in "${data[@]}"; do
    echo "Closing charge num: ${m}..."
    sudo docker exec  teslamate_teslamate_1 /opt/app/bin/teslamate rpc "TeslaMate.Repo.get(TeslaMate.Log.ChargingProcess,${m}) |> TeslaMate.Log.complete_charging_process()";
  done
fi

sql="select id from drives where end_date is null and start_date <  now() + interval '-2 days' order by start_date;"
data=$(docker exec teslamate_database_1 bash -c "psql -d teslamate -U teslamate -t -c \"${sql}\"")
data=($data)

if [ "${data}" != "" ]; then
  for m in "${data[@]}"; do
    echo "Closing drive num: ${m}..."
    sudo docker exec  teslamate_teslamate_1 /opt/app/bin/teslamate rpc "TeslaMate.Repo.get(TeslaMate.Log.Drive,${m}) |> TeslaMate.Log.close_drive()";
  done
fi


echo "Done."
```

**Reply by cwanja** [reply]
@dmachadol this worked perfectly with a minor modification. I had to update `teslamate_database_1` and `teslamate_teslamate_1` to `teslamate-database-1` and `teslamate-teslamate-1` respectively. Going to add this script to our documentation. 

Just closed over 270 drives thanks to @jheredianet dashboard. PS Juan, can you make a PR to add the incomplete data dashboard? Alternative could be to add a panel to _both_ **Drive and Charge Stats** with a count of incomplete objects. We can let a discussion in the PR take place. Tag me in the PR so I can track it please 😄 thanks!

**Reply by cwanja** [reply]
Docs update in PR #3877

**Comment by PavaVo** [comment]
I've implemented Teslamate on Unraid 7.2.x (yeah I know: it's not official supported, but got it working anyway). I needed to close more than 1050 drives and 1400 charges. So, some automation was very welcome! 

Build some more scripts for Teslamate on Unraid; need to find a place to publish them for the greater good; any suggestions?

```shell
#!/usr/bin/env bash
set -euo pipefail

DB_CONTAINER="postgresql17"
TM_CONTAINER="teslamate"
DB_NAME="teslamate"
DB_USER="teslamate"

echo "=== TeslaMate Batch Fix (Unraid) ==="

echo "Fetching incomplete drives..."
drive_ids=$(docker exec -i "$DB_CONTAINER" psql \
  -U "$DB_USER" -d "$DB_NAME" -t -c \
  "SELECT id FROM drives WHERE end_date IS NULL;" | sed '/^\s*$/d')

echo "Fetching incomplete charging processes..."
charge_ids=$(docker exec -i "$DB_CONTAINER" psql \
  -U "$DB_USER" -d "$DB_NAME" -t -c \
  "SELECT id FROM charging_processes WHERE end_date IS NULL;" | sed '/^\s*$/d')
echo "Drives to fix:   $(echo "$drive_ids" | wc -l)"
echo "Charges to fix:  $(echo "$charge_ids" | wc -l)"

for id in $drive_ids; do
  echo "Closing drive ID $id"
  docker exec -i "$TM_CONTAINER" bin/teslamate rpc \
    "TeslaMate.Repo.get!(TeslaMate.Log.Drive, $id) |> TeslaMate.Log.close_drive()"
done

for id in $charge_ids; do
  echo "Closing charge ID $id"
  docker exec -i "$TM_CONTAINER" bin/teslamate rpc \
    "TeslaMate.Repo.get!(TeslaMate.Log.ChargingProcess, $id) |> TeslaMate.Log.complete_charging_process()"
done

echo "=== TeslaMate batch repair completed successfully ==="
```
